core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -0,0 +1,430 @@
+        var currOffset = log.logStartOffset
+        while (currOffset < highWaterMark && !shuttingDown.get()) {
+          buffer.clear()
+          val fileRecords = log.read(currOffset, config.transactionLogLoadBufferSize, maxOffset = None, minOneMessage = true)
+            .records.asInstanceOf[FileRecords]
+          val bufferRead = fileRecords.readInto(buffer, 0)
+
+          MemoryRecords.readableRecords(bufferRead).batches.asScala.foreach { batch =>
+            for (record <- batch.asScala) {
+              require(record.hasKey, "Transaction state log's key should not be null")
+              TransactionLog.readMessageKey(record.key) match {
+
+                case txnKey: TxnKey =>
+                  // load transaction metadata along with transaction state
+                  val transactionalId: String = txnKey.transactionalId
+                  if (!record.hasValue) {
+                    loadedTransactions.remove(transactionalId)
+                    removedTransactionalIds.add(transactionalId)
+                  } else {
+                    val txnMetadata = TransactionLog.readMessageValue(record.value)
+                    loadedTransactions.put(transactionalId, txnMetadata)
+                    removedTransactionalIds.remove(transactionalId)
+                  }
+
+                case unknownKey =>
+                  throw new IllegalStateException(s"Unexpected message key $unknownKey while loading offsets and group metadata")
+              }
+
+              currOffset = batch.nextOffset
+            }
+          }
+
+          loadedTransactions.foreach {
+            case (transactionalId, txnMetadata) =>
+              val currentTxnMetadata = addTransaction(transactionalId, txnMetadata)
+              if (!txnMetadata.equals(currentTxnMetadata)) {
+                // treat this as a fatal failure as this should never happen
+                fatal(s"Attempt to load $transactionalId's metadata $txnMetadata failed " +
+                  s"because there is already a different cached transaction metadata $currentTxnMetadata.")
+
+                throw new KafkaException("Loading transaction topic partition failed.")
+              }
+          }
+
+          removedTransactionalIds.foreach { transactionalId =>
+            if (transactionMetadataCache.contains(transactionalId)) {
+              // the cache already contains a transaction which should be removed,
+              // treat this as a fatal failure as this should never happen
+              fatal(s"Unexpected to see $transactionalId's metadata while " +
+                s"loading partition $topicPartition since its latest state is a tombstone")
+
+              throw new KafkaException("Loading transaction topic partition failed.")
+            }
+          }
+
+          info(s"Finished loading ${loadedTransactions.size} transaction metadata from $topicPartition in ${time.milliseconds() - startMs} milliseconds")
```

Comment:
Yes. We need to do that. Discussed this with @guozhangwang offline yesterday and we thought it might be better to do in a follow up patch as this patch is already quite large. Thoughts?